### PR TITLE
Basic functionality for running a DGE experiment.

### PR DESCRIPTION
**Pull request by apfejes**
This is a first pass at basic functionality - and has been used for running DGE on real data with good looking results. 

However, a full review would help this branch improve a lot.

### CODE REVIEW

File: requirements-test.txt [hunk 1]
```diff
@@ -7,4 +7,5 @@ pytest-cov==2.5.1
 pytest-parallel==0.0.2
 pymongo==3.7.1
 pytest-mongodb==2.1.2
-black==18.6b3
\ No newline at end of file
+black==18.6b3
```
Comment (Emiller88) [inline requirements-test.txt]: Any way to clean up these two lines?

File: scripts/edgepy.py [hunk 2]
```diff
@@ -66,15 +111,115 @@ def __init__(self, args):
             # Todo: convert this to a dge_list object.
             # self.dge_list = ???
 
-    def run(self):
-        pass
+        self.output = args.output if args.output else None
+        self.p_value_cutoff = args.cutoff
+        self.minimum_cpm = args.minimum_cpm
```
Comment (Emiller88) [inline scripts/edgepy.py]: Indention might be off here.

File: requirements-test.txt [hunk 3]
```diff
@@ -7,4 +7,5 @@ pytest-cov==2.5.1
...
+scipy==1.1.0
```
Comment (clintval) [inline requirements-test.txt]: scipy is not a test requirment, it should go into `install_requires` of `setup.py` FIY

File: tests/test_DGEList.py [hunk 4]
```diff
@@ -74,16 +73,16 @@ def test_library_size():
...
+                78565885,
+                48667119,
+                62799917,
+                66032107,
+                66194776,
+                55085875,
+                37760315,
             ]
         ),
     )
```
Comment (sdhutchins) [inline tests/test_DGEList.py]: Inconsistent np.array structure

File: edgePy/DGEList.py [hunk 5]
```diff
@@ -28,17 +28,23 @@ class DGEList(object):
...
     """
 
     # Pattern to delete from field names anytime they are assigned.
     _field_strip_re = re.compile(r'[\s"]+')
 
+    # Metatags used in older HTSeq datasets without underscore prefixes.
+    _old_metatags = np.array(
+        ['no_feature', 'ambiguous', 'too_low_aQual', 'not_aligned', 'alignment_not_unique']
+    )
+
```
Comment (sdhutchins) [inline edgePy/DGEList.py]: Inconsistent np.array structure compared to in tests/test_DGEList.py

File: edgePy/DGEList.py [hunk 6]
```diff
@@ -353,22 +365,40 @@ def create_DGEList(
...
-    def create_DGEList_handle(
-        cls, data_handle: StringIO, group_file: Path, **kwargs: Mapping
+    def create_DGEList_data_file(
+        cls, data_file: Path, group_file: Path, **kwargs: Mapping
     ) -> "DGEList":
         """Read in a file-like object of delimited data for instantiation.
 
         Args:
-            data_handle: Any handle supporting text streaming io.
+            data_file: Text File defining the data set.
```
Comment (sdhutchins) [inline edgePy/DGEList.py]: `File` should be lowercase.

File: edgePy/DGEList.py [hunk 7]
```diff
@@ -353,22 +365,40 @@ def create_DGEList(
...
             group_file: the json file defining the groups
```
Comment (sdhutchins) [inline edgePy/DGEList.py]: `the` should probably be uppercase. Otherwise docstrings look great!

File: requirements-test.txt (hunk 1 above)
Comment (apfejes) [reply requirements-test.txt]: Sorry - what is it you'd like cleaned up?

File: requirements-test.txt (hunk 3 above)
Comment (apfejes) [reply requirements-test.txt]: done.

File: tests/test_DGEList.py (hunk 4 above)
Comment (apfejes) [reply tests/test_DGEList.py]: This is formatting from our automated code formatter, so nothing to do, unfortunately.

File: edgePy/DGEList.py (hunk 6 above)
Comment (apfejes) [reply edgePy/DGEList.py]: fixed

File: edgePy/DGEList.py (hunk 5 above)
Comment (apfejes) [reply edgePy/DGEList.py]: Can't fix - this is how black does the format.

File: edgePy/DGEList.py (hunk 7 above)
Comment (apfejes) [reply edgePy/DGEList.py]: fixed.

File: requirements-test.txt (hunk 1 above)
Comment (clintval) [reply requirements-test.txt]: Hey's just asking why there is a `diff` here. As in, what did you change?

File: scripts/edgepy.py [hunk 8]
```diff
@@ -21,6 +28,21 @@ def parse_arguments(parser=None):
     )
     parser.add_argument("--mongo_key_name", default="Project")
     parser.add_argument("--mongo_key_value", default="RNA-Seq1")
+    parser.add_argument(
+        "--group1_sample_names", nargs='+', help="L:ist of samples names for first group"
+    )
+    parser.add_argument(
+        "--group2_sample_names", nargs='+', help="L:ist of samples names for second group"
```
Comment (clintval) [inline scripts/edgepy.py]: `L:ist`?

File: scripts/edgepy.py [hunk 9]
```diff
@@ -21,6 +28,21 @@ def parse_arguments(parser=None):
     )
     parser.add_argument("--mongo_key_name", default="Project")
     parser.add_argument("--mongo_key_value", default="RNA-Seq1")
+    parser.add_argument(
+        "--group1_sample_names", nargs='+', help="L:ist of samples names for first group"
```
Comment (clintval) [inline scripts/edgepy.py]: `L:ist`?

File: scripts/edgepy.py [hunk 10]
```diff
@@ -37,39 +59,177 @@ def __init__(self, args):
...
+            else:
+                raise ValueError("Insufficient parameters for use of Mongodb")
+
             mongo_importer = ImportFromMongodb(
                 host=config.get("Mongo", "host"),
                 port=config.get("Mongo", "port"),
-                mongo_key_name=args.mongo_key_name,
-                mongo_key_value=args.mongo_key_value,
+                mongo_key_name=key_name,
+                mongo_key_value=key_value,
```
Comment (clintval) [inline scripts/edgepy.py]: Why `key_name` and `key_value` and not just `key` and `value` like how people think about dictionaries?

File: scripts/edgepy.py [hunk 11]
```diff
@@ -37,39 +59,177 @@ def __init__(self, args):
...
-    def run(self):
-        pass
+        self.output = args.output if args.output else None
+        self.p_value_cutoff = args.cutoff
+        self.minimum_cpm = args.minimum_cpm
+
+    def fake_groups(self):
+        """
+        This is used to create fake groups, if you didn't give any at the time the DGEList object was
+         created... this is a bug that needs to be fixed.
```
Comment (clintval) [inline scripts/edgepy.py]: Can we get a unit test with an `pytest.mark.xfail` on it that will mark we need to eventually fix this?

File: scripts/edgepy.py [hunk 12]
```diff
@@ -37,39 +59,177 @@ def __init__(self, args):
...
+        :return:
+        """
+        self.dge_list.group = [
+            'fibroblast' if 'fibroblast' in g else 'cardiomyocyte' for g in self.dge_list.group
```
Comment (clintval) [inline scripts/edgepy.py]: Do not hardcode. Consider defining project defaults at top of scripts like:


```python
DEFAULT_TEST_TISSUE = 'fibroblast'
```

Then use it later.

File: scripts/edgepy.py [hunk 13]
```diff
@@ -37,39 +59,177 @@ def __init__(self, args):
...
+    def run_ks(self):
+        """
+        First pass implementation of a Kolmogorov-Smirnov test for different groups, using the Scipy KS test two-tailed
+        implementation.
+
+        Args:
+             None.
+        """
+
+        print(self.dge_list.groups_list)
```
Comment (clintval) [inline scripts/edgepy.py]: Use `logging` and set to `DEBUG`. I want to see all `print` statements converted to logs.

File: scripts/edgepy.py [hunk 14]
```diff
@@ -37,39 +59,177 @@ def __init__(self, args):
...
+
+        gene_details, gene_likelyhood1, group_types = self.ks_2_samples()
```
Comment (clintval) [inline scripts/edgepy.py]: sp? likelihood.

Comment (clintval) [reply scripts/edgepy.py]: This misspelling appears elsewhere too FYI

File: scripts/edgepy.py [hunk 15]
```diff
@@ -37,39 +59,177 @@ def __init__(self, args):
...
+        if self.output:
+            with smart_open(self.output, 'w') as out:
+                out.writelines(results)
+            print(f"wrote to {self.output}")
+        else:
+            for line in results:
+                print(line)
+
+    def ks_2_samples(self):
+        gene_likelyhood1: Dict[Hashable, float] = {}
```
Comment (clintval) [inline scripts/edgepy.py]: No `docstring` on what this function does? It's not apparent to me.

File: setup.cfg [hunk 16]
```diff
@@ -27,6 +27,7 @@ install_requires =
     numpy>=1.14.5
     smart_open>=1.6.0
     tox>=3.1.2
+    scipy>=1.1.0
```
Comment (clintval) [inline setup.cfg]: Excellent!

File: requirements-test.txt (hunk 1 above)
Comment (apfejes) [reply requirements-test.txt]: Ah, this was just adding black as a dependency for testing.  Going to mark this as resolved, as there's nothing to do here.

File: edgePy/DGEList.py [hunk 17]
```diff
@@ -340,27 +340,27 @@ def create_DGEList(
...
         data_set: Dict[Hashable, Any],  # {sample: {gene1: x, gene2: y}},
         gene_list: List[str],
-        sample_category: Dict[Hashable, str],
+        sample_category_list: List[str] = None,
+        sample_category_dict: Dict[Hashable, List[str]] = None,
     ) -> "DGEList":
         """ sample list and gene list must be pre-sorted
             Use this to create the DGE object for future work."""
 
         print("Creating DGE list object...")
```
[the remark this answers is not in the document]
Comment (apfejes) [reply edgePy/DGEList.py]: Created this as an issue.  It's too big to do in this change, but entirely appropriate!

File: edgePy/DGEList.py [hunk 18]
```diff
@@ -340,27 +340,27 @@ def create_DGEList(
         sample_list: List[str],
         data_set: Dict[Hashable, Any],  # {sample: {gene1: x, gene2: y}},
         gene_list: List[str],
-        sample_category: Dict[Hashable, str],
+        sample_category_list: List[str] = None,
+        sample_category_dict: Dict[Hashable, List[str]] = None,
```
[the remark this answers is not in the document]
Comment (apfejes) [reply edgePy/DGEList.py]: Fixed.

File: edgePy/DGEList.py [hunk 19]
```diff
@@ -371,8 +371,8 @@ def create_DGEList_data_file(
         """Read in a file-like object of delimited data for instantiation.
 
         Args:
-            data_file: Text File defining the data set.
-            group_file: the json file defining the groups
+            data_file: Text file defining the data set.
+            group_file: The json file defining the groups.
```
[the remark this answers is not in the document]
Comment (apfejes) [reply edgePy/DGEList.py]: Fixed.

File: edgePy/DGEList.py [hunk 20]
```diff
@@ -340,27 +340,27 @@ def create_DGEList(
         sample_list: List[str],
         data_set: Dict[Hashable, Any],  # {sample: {gene1: x, gene2: y}},
         gene_list: List[str],
-        sample_category: Dict[Hashable, str],
+        sample_category_list: List[str] = None,
```
[the remark this answers is not in the document]
Comment (apfejes) [reply edgePy/DGEList.py]: I was doing this to make it clear which one is which - there are two ways in which categories are needed, by sample (eg, sample -> category,  as a list) and by category (category -> [samples], as a dict).  I used the names to help clean up the code base, but that's not stopping me from fixing it now.... 

I will refactor to better names.

File: scripts/edgepy.py (hunk 10 above)
Comment (apfejes) [reply scripts/edgepy.py]: thanks - good point!

File: scripts/edgepy.py (hunk 13 above)
Comment (apfejes) [reply scripts/edgepy.py]: already filed an issue for this.  Will tackle next.

File: scripts/edgepy.py (hunk 14 above)
Comment (apfejes) [reply scripts/edgepy.py]: fixed.

File: edgePy/data_import/mongodb/mongo_import.py [hunk 21]
```diff
@@ -81,12 +82,15 @@ def translate_gene_list(self, database: str) -> None:
             self.gene_list = ensg_genes
 
     def get_data_from_mongo(
-        self, database: str = "Tenaya"
+        self, database: str = "Tenaya", rpkm_flag: bool = False
```
[the remark this answers is not in the document]
Comment (apfejes) [reply edgePy/data_import/mongodb/mongo_import.py]: ugh - forgot to take it out.  Thanks for catching it.

File: scripts/edgepy.py (hunk 11 above)
Comment (apfejes) [reply scripts/edgepy.py]: I believe I've already addressed the issue this was fixing. I'll remove the function.

File: scripts/edgepy.py (hunk 12 above)
Comment (apfejes) [reply scripts/edgepy.py]: code has been removed.

File: scripts/edgepy.py (hunk 15 above)
Comment (apfejes) [reply scripts/edgepy.py]: docstring added.

File: edgePy/DGEList.py [hunk 22]
```diff
@@ -340,27 +340,27 @@ def create_DGEList(
...
     ) -> "DGEList":
         """ sample list and gene list must be pre-sorted
             Use this to create the DGE object for future work."""
 
         print("Creating DGE list object...")
         temp_data_store = np.zeros(shape=(len(gene_list), len(sample_list)))
-        group = []
 
         for idx_s, sample in enumerate(sample_list):
             for idx_g, gene in enumerate(gene_list):
```
[the remark this answers is not in the document]
Comment (apfejes) [reply edgePy/DGEList.py]: it's just converting the dictionary to a numpy 2d array.  If there's a better way to do that, let me know.

**Pull request merged**